### PR TITLE
Adopting pathlib for main.py

### PR DESCRIPTION
**Pull request by cakester**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

Phased refactor adoption of loader/segmentation_pair.py as detailed in #872 using pathlib for using pathlib for using pathlib for main.py

## Linked issues

Step wise remediation of #872

### CODE REVIEW

File: ivadomed/main.py [hunk 1]
```diff
@@ -72,15 +72,15 @@ def get_parser():
 
 
 def create_path_model(context, model_params, ds_train, path_output, train_onehotencoder):
-    path_model = os.path.join(path_output, context["model_name"])
-    if not os.path.isdir(path_model):
+    path_model = Path(path_output, context["model_name"])
```
Comment (dyt811) [inline ivadomed/main.py:75]: best to use `get()` pattern and potentially convert that string to utilize the `keyword.py`

File: ivadomed/main.py [hunk 2]
```diff
@@ -72,15 +72,15 @@ def get_parser():
...
+    if not path_model.is_dir():
         logger.info('Creating model directory: {}'.format(path_model))
```
Comment (dyt811) [inline ivadomed/main.py:77]: `f` string adoption wherever we can would be good going forward.

File: ivadomed/main.py [hunk 3]
```diff
@@ -72,15 +72,15 @@ def get_parser():
...
         if 'film_layers' in model_params and any(model_params['film_layers']):
-            joblib.dump(train_onehotencoder, os.path.join(path_model, "one_hot_encoder.joblib"))
+            joblib.dump(train_onehotencoder, path_model.joinpath("one_hot_encoder.joblib"))
             if 'metadata_dict' in ds_train[0]['input_metadata'][0]:
                 metadata_dict = ds_train[0]['input_metadata'][0]['metadata_dict']
-                joblib.dump(metadata_dict, os.path.join(path_model, "metadata_dict.joblib"))
+                joblib.dump(metadata_dict, path_model.joinpath("metadata_dict.joblib"))
 
     else:
         logger.info('Model directory already exists: {}'.format(path_model))
```
Comment (dyt811) [inline ivadomed/main.py:86]: `f`

File: ivadomed/main.py [hunk 4]
```diff
@@ -107,8 +107,8 @@ def film_normalize_data(context, model_params, ds_train, ds_valid, path_output):
                                             model_params['metadata'])
     model_params.update({"film_onehotencoder": train_onehotencoder,
                          "n_metadata": len([ll for l in train_onehotencoder.categories_ for ll in l])})
-    joblib.dump(metadata_clustering_models, os.path.join(path_output, "clustering_models.joblib"))
-    joblib.dump(train_onehotencoder, os.path.join(path_output + "one_hot_encoder.joblib"))
+    joblib.dump(metadata_clustering_models, Path(path_output, "clustering_models.joblib"))
+    joblib.dump(train_onehotencoder, Path(path_output + "one_hot_encoder.joblib"))
```
Comment (dyt811) [inline ivadomed/main.py:111]: Nice. TIL `joblib.dump` can provide handle both str of Path

File: ivadomed/main.py [hunk 5]
```diff
@@ -124,9 +124,9 @@ def get_dataset(bids_df, loader_params, data_lst, transform_params, cuda_availab
 def save_config_file(context, path_output):
     # Save config file within path_output and path_output/model_name
     # Done after the threshold_analysis to propate this info in the config files
-    with open(os.path.join(path_output, "config_file.json"), 'w') as fp:
+    with Path(path_output, "config_file.json").open(mode='w') as fp:
         json.dump(context, fp, indent=4)
-    with open(os.path.join(path_output, context["model_name"], context["model_name"] + ".json"), 'w') as fp:
+    with Path(path_output, context["model_name"], context["model_name"] + ".json").open(mode='w') as fp:
```
Comment (dyt811) [inline ivadomed/main.py:129]: Please ensure to adjust any future instances to use `.get()`  method to access dictionary. We really need to remove this antipattern.

File: ivadomed/main.py [hunk 6]
```diff
@@ -190,20 +190,20 @@ def set_model_params(context, loader_params):
...
     return path_output
 
 
 def update_film_model_params(context, ds_test, model_params, path_output):
-    clustering_path = os.path.join(path_output, "clustering_models.joblib")
+    clustering_path = Path(path_output, "clustering_models.joblib")
     metadata_clustering_models = joblib.load(clustering_path)
     # Model directory
-    ohe_path = os.path.join(path_output, context["model_name"], "one_hot_encoder.joblib")
+    ohe_path = Path(path_output, context["model_name"], "one_hot_encoder.joblib")
```
Comment (dyt811) [inline ivadomed/main.py:206]: `get()` antipattern removal

File: ivadomed/main.py [hunk 7]
```diff
@@ -222,11 +222,11 @@ def run_segment_command(context, model_params):
     bids_subjects = sorted(bids_df.df['filename'].to_list())
 
     # Add postprocessing to packaged model
-    path_model = os.path.join(context['path_output'], context['model_name'])
-    path_model_config = os.path.join(path_model, context['model_name'] + ".json")
-    model_config = imed_config_manager.load_json(path_model_config)
+    path_model = Path(context['path_output'], context['model_name'])
+    path_model_config = Path(path_model, context['model_name'] + ".json")
```
Comment (dyt811) [inline ivadomed/main.py:226]: `get()`

File: ivadomed/main.py [hunk 8]
```diff
@@ -222,11 +222,11 @@ def run_segment_command(context, model_params):
...
+    model_config = imed_config_manager.load_json(str(path_model_config))
     model_config['postprocessing'] = context['postprocessing']
```
Comment (dyt811) [inline ivadomed/main.py:228]: this line too.

File: ivadomed/main.py [hunk 9]
```diff
@@ -271,24 +271,24 @@ def run_segment_command(context, model_params):
...
         if fname_img:
-            pred_list, target_list = imed_inference.segment_volume(path_model,
+            pred_list, target_list = imed_inference.segment_volume(str(path_model),
                                                                    fname_images=fname_img,
                                                                    gpu_id=context['gpu_ids'][0],
                                                                    options=options)
-            pred_path = os.path.join(context['path_output'], "pred_masks")
-            if not os.path.exists(pred_path):
-                os.makedirs(pred_path)
+            pred_path = Path(context['path_output'], "pred_masks")
```
Comment (dyt811) [inline ivadomed/main.py:278]: `.get()`

File: ivadomed/main.py [hunk 10]
```diff
@@ -271,24 +271,24 @@ def run_segment_command(context, model_params):
...
+            if not pred_path.exists():
+                pred_path.mkdir(parents=True)
 
             for pred, target in zip(pred_list, target_list):
                 filename = subject.split('.')[0] + target + "_pred" + ".nii.gz"
-                nib.save(pred, os.path.join(pred_path, filename))
+                nib.save(pred, Path(pred_path, filename))
```
Comment (dyt811) [inline ivadomed/main.py:284]: Good point that `nib` can save both via `os` or `path`!

File: ivadomed/main.py [hunk 11]
```diff
@@ -319,9 +319,9 @@ def run_command(context, n_gif=0, thr_increment=None, resume_training=False):
     """
     command = copy.deepcopy(context["command"])
     path_output = set_output_path(context)
-    log_file = os.path.join(context['path_output'], context['log_file'])
+    log_file = Path(context['path_output'], context['log_file'])
```
Comment (dyt811) [inline ivadomed/main.py:322]: `.get()` patch here
also recommend to change variable name to `path_log`

File: ivadomed/main.py [hunk 12]
```diff
@@ -495,10 +495,10 @@ def create_dataset_and_ivadomed_version_log(context):
         for Dataset in path_data:
             datasets_version.append(imed_utils.__get_commit(path_to_git_folder=Dataset))
 
-    log_file = os.path.join(context['path_output'], 'version_info.log')
+    log_file = Path(context['path_output'], 'version_info.log')
```
Comment (dyt811) [inline ivadomed/main.py:498]: `path_log` renaming recommended. and also `.get()`

**Review by dyt811 — CHANGES_REQUESTED** [review]
Minor few issues not with current PR but in the regions we touched are strongly recommend to be updated and bring to the latest standard to avoid potential exceptions.

File: ivadomed/main.py (hunk 1 above)
Comment (cakester) [reply ivadomed/main.py:75]: keyword.py changes will be in another pr

**Review by dyt811 — APPROVED** [review]
Thanks for correcting everything so promptly. All existing changes looking good. Please resolve the conflicts carefully and it should be good to merge.

**Pull request merged**